iroh-gossip/src/proto/state.rs

```diff
@@ -0,0 +1,253 @@
+    }
+
+    pub fn state_mut(&mut self, topic: &TopicId) -> Option<&mut topic::State<PA, R>> {
+        self.states.get_mut(topic)
+    }
+
+    pub fn has_active_peers(&self, topic: &TopicId) -> bool {
+        self.state(topic)
+            .map(|s| s.has_active_peers())
+            .unwrap_or(false)
+    }
+
+    pub fn handle(
+        &mut self,
+        event: InEvent<PA>,
+        now: Instant,
+    ) -> impl Iterator<Item = OutEvent<PA>> + '_ {
+        let event: InEventMapped<PA> = event.into();
+
+        // todo: add command to leave a topic
+        match event {
+            InEventMapped::TopicEvent(topic, event) => {
+                // when receiving messages, update our conn map to take note that this topic state may want
+                // to keep this connection
+                // TODO: this is a lot of hashmap lookups, maybe there's ways to optimize?
+                if let topic::InEvent::RecvMessage(from, _message) = &event {
+                    self.conns.entry(*from).or_default().insert(topic);
+                }
+                // when receiving a join command, initialize state if it doesn't exist
+                if let topic::InEvent::Command(topic::Command::Join(_peer)) = &event {
+                    if let hash_map::Entry::Vacant(e) = self.states.entry(topic) {
+                        e.insert(topic::State::with_rng(
+                            self.me,
+                            self.me_data.clone(),
+                            self.config.clone(),
+                            self.rng.clone(),
+                        ));
+                    }
+                }
+                if let hash_map::Entry::Vacant(e) = self.states.entry(topic) {
+                    e.insert(topic::State::with_rng(
+                        self.me,
+                        self.me_data.clone(),
+                        self.config.clone(),
+                        self.rng.clone(),
+                    ));
+                }
+
+                // pass the event to the state handler
+                if let Some(state) = self.states.get_mut(&topic) {
+                    let out = state.handle(event, now);
+                    for event in out {
+                        handle_out_event(topic, event, &mut self.conns, &mut self.outbox);
+                    }
+                }
+            }
+            // when a peer disconnected on the network level, forward event to all states
+            InEventMapped::All(event) => {
+                if let topic::InEvent::UpdatePeerData(data) = &event {
+                    self.me_data = data.clone();
+                }
+                for (topic, state) in self.states.iter_mut() {
+                    let out = state.handle(event.clone(), now);
+                    for event in out {
+                        handle_out_event(*topic, event, &mut self.conns, &mut self.outbox);
+                    }
+                }
+            }
+        }
+
+        self.outbox.drain(..)
```

Comment:
what exactly is the point of this?

Reply:
So the `State` assembles a list of `OutEvent`s and then returns an `Iterator` of those `OutEvent`s to the caller. 
The though was: a) keep it an implementation detail how the outevents are stored in the state inbetween, and b) possibly reduce allocations, as the `Vec` that stores the outevents in the state is reused this way, and the caller usually can handle the events right while iterating.

Does that answer your question? Or, what would you suggest instead?